src/Shared/BuildEnvironmentHelper.cs

```diff
@@ -126,190 +93,204 @@ private static BuildEnvironment Initialize()
-            // is found.
-            var possibleLocations = new Func<string>[]
+            if (instances.Count == 0) return null;
+
+            if (instances.Count > 1)
             {
-                () => TryGetVsFromProcess(processNameCommandLine),
-                () => TryGetVsFromProcess(processNameCurrentProcess),
-                () => TryGetVsFromEnvironment(),
-                () => TryGetVsFromInstalled(),
-                () => TryGetVsFromMSBuildLocation(processNameCommandLine),
-                () => TryGetVsFromMSBuildLocation(processNameCurrentProcess)
-            };
+                // TODO: Warn user somehow. We may have picked the wrong one.
+            }
 
-            return possibleLocations.Select(location => location()).FirstOrDefault(path => !string.IsNullOrEmpty(path));
+            return new BuildEnvironment(
+                BuildEnvironmentMode.VisualStudio,
+                GetMsBuildExeFromVsRoot(instances[0].Path),
+                runningTests: false,
+                runningInVisualStudio: false,
+                visualStudioPath: instances[0].Path);
         }
 
-        private static string TryGetVsFromProcess(string process)
+        private static BuildEnvironment TryFromAppContextBaseDirectory()
         {
-            // Check to see if we're running inside of Visual Studio
-            // This assumes running from VS\Common7\IDE\<process>.exe.
-            return IsProcessInList(process, s_visualStudioProcess)
-                ? FileUtilities.GetFolderAbove(process, 3)
-                : null;
+            // Assemblies compiled against anything older than .NET 4.0 won't have a System.AppContext
+            // Try the base directory that the assembly resolver uses to probe for assemblies.
+            // Under certain scenarios the assemblies are loaded from spurious locations like the NuGet package cache
+            // but the toolset files are copied to the app's directory via "contentFiles".
+            return TryFromStandaloneMsBuildExe(s_getAppContextBaseDirectory());
         }
 
-        private static string TryGetVsFromEnvironment()
+        private static BuildEnvironment TryFromStandaloneMsBuildExe(string msBuildExePath)
         {
-            // VSInstallDir is set from the Developer Command Prompt
-            var vsInstallDir = Environment.GetEnvironmentVariable("VSINSTALLDIR");
-            var vsVersion = Environment.GetEnvironmentVariable("VisualStudioVersion");
-
-            if (!string.IsNullOrEmpty(vsInstallDir) &&
-                !string.IsNullOrEmpty(vsVersion) &&
-                vsVersion == CurrentVisualStudioVersion &&
-                Directory.Exists(vsInstallDir))
+            if (!string.IsNullOrEmpty(msBuildExePath) && File.Exists(msBuildExePath))
             {
-                return vsInstallDir;
+                // %MSBUILD_EXE_PATH% was defined. Assume 'Standalone' mode.
```

Comment:
Is this comment valid? It refers to an environment variable but the exe comes in as a parameter. Related to my previous comment on unifying the MSBuildFromVS and MSBuildStandalone cases